### PR TITLE
[WIP] Add LinearAlgebra::SharedMPI::Vector

### PR DESCRIPTION
**Pull request by peterrum**
### LinearAlgebra::SharedMPI::Vector

This PR adds a new vector `LinearAlgebra::SharedMPI::Vector` class that is built around MPI 3.0 shared-memory features.

MPI 3.0 provides functions to allocate memory for an array with the following command:
```cpp
MPI_Win_allocate_shared(size, sizeof(Number), info, comm_shared, &data_this, win);
```
and to query a pointer to the beginning of the array owned by processes on the shared-memory domain: 
```cpp
MPI_Win_shared_query(*win, i, &ssize, &disp_unit, &data.others[i]);
```

The command to create a communicator `comm_shared` is:
```cpp
MPI_Comm_split_type(
        comm, MPI_COMM_TYPE_SHARED, rank, MPI_INFO_NULL, &comm_shared);
```

These functions allow us to access all vector entries on the same compute node (same shared-memory domain) also in a purely MPI-parallelized program.  

For more implementation details, see Section 4 and Appendix A of the pre-print of the hyper.deal release paper (https://arxiv.org/abs/2002.08110).

### LinearAlgebra::SharedMPI::PartitionerBase

A second feature of the new vector class is that users can pass their own partitioner implementation.  The first implementation `LinearAlgebra::SharedMPI::Partitioner` is just like `Utilities::MPI::Partitioner` (built around `IndexSet`) but splits up (internally) `export_to_ghost_array` and `import_from_ghost_array` in two steps: one for remote data (which requires `MPI_Send/MPI_Recv`) and one for shared data (which copied via `memcpy` if buffering is requested).

The interface should look familiar to that of `Utilities::MPI::Partitioner` with the difference that we are working with vectors of pointers (which are returned by `MPI_Win_shared_query`).

In hyper.deal, we are using our own partitioner implementation, which is specially tailored for DG. 

### Motivation

In hyper.deal, we reached a speed up of 25-30% on a single compute node (with ECL) by not using `MPI_Send/MPI_Recv` and in CEED BP1, a speed up up to 10% for the operator evaluation. Our hope is to reach better results with DG.

### Next steps

In the next step, we will extend the support of `MatrixFree` for the new vector class, in particular, so that we can use it for DG. Furthermore, we would like to investigate partitioner strategies within `MatrixFree`, which requires some internal changes - so let's keep that out of this PR!

*Note: I still need to clean up the code and add some documentation. Nevertheless, a preliminary feedback would be push appretated! Maybe someone has suggestions how to integrate the new classes better in the existing deal.II concepts.*

Related to https://github.com/hyperdeal/hyperdeal/pull/18 and https://github.com/kronbichler/ceed_benchmarks_dealii/pull/7.

### CODE REVIEW

File: include/deal.II/matrix_free/matrix_free.h
Comment (kronbichler) [inline include/deal.II/matrix_free/matrix_free.h:3783]: I have not looked into all details, but generally `reset_ghost_values` is for the vector we called `update_ghost_values` on (but which we do not want to remain ghosted afterwards).

**Review by kronbichler — COMMENTED** [review]
I have collected some random comments. I agree with the basic concepts because I have been discussing them with @peterrum before, but it would be good with some additional comments.

File: include/deal.II/lac/la_sm_partitioner.h
Comment (peterrum) [reply include/deal.II/lac/la_sm_partitioner.h:214]: The members are part of the base class because they are needed by some methods in the base class. I don't think these methods need the be implemented by the implementations.

File: include/deal.II/lac/la_sm_partitioner.h
Comment (peterrum) [reply include/deal.II/lac/la_sm_partitioner.h:34]: What do you suggest? To move all the partitioners here?

File: include/deal.II/lac/la_sm_partitioner.h
Comment (peterrum) [reply include/deal.II/lac/la_sm_partitioner.h:196]: The annoying thing is that we the method is called the same way. Any other name suggestions?

Comment (bangerth) [reply include/deal.II/lac/la_sm_partitioner.h:196]: `number_of_mpi_processes`

File: include/deal.II/lac/la_sm_vector.templates.h
Comment (peterrum) [reply include/deal.II/lac/la_sm_vector.templates.h:76]: I have fixed this. According to the source code of Open MPI
```cpp
    *((void**) baseptr) = base;
```
see also https://github.com/open-mpi/ompi/blob/6c46da32454553a52c6b0c30cae8d0075c43cd94/ompi/win/win.c#L323, the function `MPI_Win_allocate_shared` expects a `void**`, i.e, in our case `&data_this`...

**Comment by peterrum** [comment]
@kronbichler I have addressed many of your comments. There are some where I am not sure how to proceed.

File: include/deal.II/lac/la_sm_partitioner.h
Comment (kronbichler) [reply include/deal.II/lac/la_sm_partitioner.h:34]: I guess we can leave things here for the moment until someone feels a need to clean up things. I really like having the partitioner file to be named similarly and in the same place. For the basic partitioner that ship has sailed (well, we can always deprecate things, but I don't think it is worth it).

File: include/deal.II/lac/la_sm_partitioner.h
Comment (kronbichler) [reply include/deal.II/lac/la_sm_partitioner.h:215]: OK.

File: include/deal.II/lac/la_sm_vector.h
Comment (kronbichler) [reply include/deal.II/lac/la_sm_vector.h:656]: The class does some checks depending on where the type traits of the vector send down the evaluator. I do not feel strongly here, so we can leave things in this state for the moment until we have all members filled.

**Review by kronbichler — APPROVED** [review]
I think this is a good step forward. Of course, we still need to fill interfaces for the vector before we can use it more generally, but I would be fine to leave that to another PR (if we manage to finalize things during the next month). It would be great if someone else can give it a look as well.

**Comment by tjhei** [comment]
I have some fundamental questions (without having looked at the code very closely):
1. To me this seems like an optimization of the existing parallel vector (to grant "local" access to entries on the same node). Is there anything else that is new/different?
2. Why would one not want to use this over the existing parallel vector?
3. If there is no good reason to do so in 2), why is this not an extension to the existing vector class? This is not particularly user-friendly to introduce another new vector type. Maybe this can be a setting/policy/compile time setting?
4. Does this mean that we need a lot of new template instantiations for the new vector type? That would worry me.
5. I am not sure I like the name "SharedMPI" (not that I have a better idea, but it reminds me of shared::Tria).

**Comment by peterrum** [comment]
> To me this seems like an optimization of the existing parallel vector (to grant "local" access to entries on the same node). Is there anything else that is new/different?

I had the option to introduce 1) a new `MemorySpace`-type  or 2) a new modus/setting of `L:d:V`  or 3)  a new vector. 

The first two options would lead to the fact that `L:d:V` would have per default information regarding shared memory: 2 communicator, functions which would only work under certain circumstances...

In my opinion, the cleanest approach is to introduce a vector, since it is different in its core. For instance, `L:d:V` is create with `IndexSet`s (which is transformed into a `Partitioner`) and such the vector has a global view on the data. However, the new vector is much simpler: it is only a container of data of a specific size (and such data can be only accessed with local indices). The interpretation of the data is completely handled by a new set of (potentially user-provided) partitioners. For instance, this PR introduces a partitioner which is built around `IndexSet`s. But in `hyper.deal` we have a tailored partitioner for `DG` (once this PR is accepted and merged, we will create a follow-up PR to move it here).

> Why would one not want to use this over the existing parallel vector?

One would need to work with a second communicator and potentially deal with race conditionally in user code.

> If there is no good reason to do so in 2), why is this not an extension to the existing vector class? This is not particularly user-friendly to introduce another new vector type. Maybe this can be a setting/policy/compile time setting?

- setting: this is more than a setting: the user needs to provide a second communicator and a partitioner (depending on the setting either one of two partitioners (old/new) and/or different methods are used)
- the policy is the partitioner, but, I don't think it is a good idea to give the old `Utilities::MPI::Partitioner` the same interface as the new one (which is specialized for the case that you have pointers to the data of other processes).
- compile-time setting has the same problems as the first point 

> Does this mean that we need a lot of new template instantiations for the new vector type? That would worry me.

The major benefit of this class will be in performance-critical code paths, i.e., `MatrixFree`: we are working on a native support for this vector class also for DG in `MatrixFree`, where we expect the largest benefit. So one could use `MatrixFree` to setup the vectors or create a `L:d:V` and/or copy over the content once from a `L:d:V`, without the need to instantiate all the functions. The first approach works very well (in `hyper.deal`) and we should probably create a set of new function in `MatrixFreeTools` (since probably all users of `MatrixFree` have implemented their own routines for computing norms, intepolating/projection some functions onto a vector, computing the right-hand side vector with `MatrixFree`).

> I am not sure I like the name "SharedMPI" (not that I have a better idea, but it reminds me of shared::Tria).

That `p:s:T` has a misleading and unfortunate name should not be a reason to dismiss the name `SharedMPI`: `p:s:T` is not shared but simply replicated on all processes, which lead to some confusion until I had a look at the implementation. The new vector is actually shared among processes via shared memory (provided by MPI).

File: include/deal.II/lac/la_sm_vector.h
Comment (tjhei) [inline include/deal.II/lac/la_sm_vector.h:40]: namespace documentation?

File: include/deal.II/lac/la_sm_vector.h
Comment (tjhei) [inline include/deal.II/lac/la_sm_vector.h:115]: uh, no documentation for the new class? Can you fix that please?

File: include/deal.II/lac/la_sm_vector.h
Comment (tjhei) [inline include/deal.II/lac/la_sm_vector.h:176]: ?

**Review by tjhei — COMMENTED** [review]
It would be great if you could write documentation for the new vector type. That would have helped with my misunderstanding what this class is trying to do.

Also, we only require MPI 2 so some of this needs to be guarded by ifdefs, right?

**Comment by tjhei** [comment]
> One would need to work with a second communicator and potentially deal with race conditionally in user code.

So maybe I misunderstood what this vector class can do. Are you saying that this is to be used only within a node (using a separate communicator) and there is no MPI/communication capabilities? An overview about what this class does would be great to have as class documentation.

**Comment by peterrum** [comment]
I am closing this PR since SM feature has been introduced into `L:p:V` in a sequence of PRs beginning from PR #11074.

**Pull request closed without merging**
